Tutorial 01: One class segmentation 2D Unet

charleygros opened the pull request:
### Context

This PR introduces a basic tutorial: training and evaluation with a basic network.

Changes include:
- New tutorial: `one_class_segmentation_2d_unet.rst`
- Various fixes to the documentation

➤ **The built documentation for this branch is available [here](https://ivadomed.org/en/cg-tutorial_oneclass/index.html#)**

### DONE

- add link to "compatible GPU is available" (section 'run the training') and point to info about compatible GPU, cuda config, etc.

### Observed issues to fix in subsequent PRs

- Issue with doc build #347
- Pytest failing for py38 #351
- Create an ivadomed download function #346
- Add a .readthedocs.yml to configure build #349
- Create PATH_IVADOMED shell environment variable #350
- Display Starting, End, Duration for training model #353
- Save log of terminal printout during training #354
- Move log_directory from config_file to ivadomed flag #355

charleygros commented:
I created a new section dedicated to Tutorials in RTD for easier search.

charleygros commented:
Note: I may put this PR on pause waiting for dataset hosted on GH repo'.

charleygros commented:
[https://ivadomed.org/en/cg-tutorial_oneclass/index.html#](https://ivadomed.org/en/cg-tutorial_oneclass/index.html#)

lrouhier commented on ivadomed/config/config.json:
```diff
@@ -1,22 +1,22 @@
     },
     "loader_parameters": {
-        "bids_path": "../duke/projects/ivadomed/spineGeneric_201907041011/result/",
+        "bids_path": "../dataset/data_spinegeneric_registered-master",
```
What about creating a folder named 'tutorial_data' that would be downloaded for some user, and therefore use this path as an example?  This way the user doesn't have to change it ?

charleygros replied:
Sorry I am not sure to understand.
Tutorial data: https://github.com/ivadomed/data_spinegeneric_registered
--> which is download by the user willing to do the tutorial by using the curl command indicated in this tutorial.

The config files in `config` are examples of configuration files for user-cases.
Currently we are also using some of them for our own experiments.. but we would like to change this: on your local machine you will have your own config file, while the ones on the repo' are more like "templates".

charleygros replied:
Thanks to Andreanne-Translator I think I better understand what you meant, sorry about that!
I agree that we should make clear in the tutorial that this path is located where the user download the data! I am going to have a try... will submit you the related doc'! Excellent point, thanks!!

charleygros replied:
Regarding the folder where they download the data: I feel they can make their own choice. I am not sure why we would want to enforce this.

charleygros replied:
Done here [0a7034d](https://github.com/ivadomed/ivadomed/pull/309/commits/0a7034de026ad2d2b1c3634abb1599ca6976507a)

lrouhier commented on docs/source/one_class_segmentation_2d_unet.rst:
```diff
@@ -0,0 +1,156 @@
+After the training for 100 epochs, the segmentations should be similar to the one presented in the following image. The output and ground truth segmentations of the spinal cord are presented in red (subject `sub-hamburg01` with contrast T2w):
+
+.. image:: ../../images/sc_prediction.png
+
```
Awesome and really clear explanation. 💪

jcohenadad commented:
oopsi!

![image](https://user-images.githubusercontent.com/2482071/87226191-2f88e980-c360-11ea-884f-a7a06aa1c6e4.png)

jcohenadad commented:
doc build failing: https://readthedocs.org/projects/ivado-medical-imaging/builds/11436005/

looking into this

jcohenadad commented:
In section 'Models available', it says that ivadomed is the only one doing that. I'm surprised. Can someone please confirm?

charleygros commented:
> In section 'Models available', it says that ivadomed is the only one doing that. I'm surprised. Can someone please confirm?

What do you mean by "doing that"? Sorry, could you please point toward the section of concern? thanks :-)

jcohenadad commented:
> > In section 'Models available', it says that ivadomed is the only one doing that. I'm surprised. Can someone please confirm?
> 
> What do you mean by "doing that"? Sorry, could you please point toward the section of concern? thanks :-)

i meant that the cell is marked as a "yes" whereas all the others have a "no". Here: 

![image](https://user-images.githubusercontent.com/2482071/87236481-3cd1c280-c3b8-11ea-8aa2-92df8e24f3af.png)

For example, a quick google search led me to this pre-trained model with monai: https://drive.google.com/drive/folders/1HXlYJGvTF3gNGOL0UFBeHVoA6Vh_GqEw

But as you can see i've added a disclaimer "we acknowledge we might be wrong, if that's the case please let us know", to dissuade frustrated researchers from starting a war with us 😉  
also, i've just renamed the column: "Pre-trained models"

jcohenadad commented on docs/source/tutorials/one_class_segmentation_2d_unet.rst:
```diff
@@ -0,0 +1,223 @@
+------------------
+
+In ``ivadomed``, training is orchestrated by a configuration file. Examples of configuration files are available in
+the ``ivadomed/config/`` and the documentation is available in :doc:`../configuration_file`.
```
this is actually a problem if people install ivadomed via pip install (which by the way is [what we recommend](https://ivadomed.org/en/cg-tutorial_oneclass/installation.html#install-from-release-recommended)), because in that case they don't have access to the `ivadomed/config`. Should we just alway point to files hosted on GitHub?  People who have cloned the repository will figure out that they also have the file locally.

jcohenadad replied:
Alternatively, given that these files are copied under the python tree when running pip install (e.g. `venv-ivadomed/lib/python3.6/site-packages/ivadomed/config/config.json`), is there an elegant way to retrieve this path locally? @Drulex or @kousu do you have an idea?

charleygros replied:
Should we change our recommendation for installation, from `pip install` to `git clone`? I feel that for this repo', a git clone should not be a problem for the potential users. What do you think @jcohenadad ?

jcohenadad replied:
I find it a shame to “downgrade” for a git clone. Why bothering with maintaining our publishing framework then? Also, pypi install is so easy! I’m sure there is a good solution to our problem here

jcohenadad commented:
@charleygros can we publish on pypi now (before mergin), so I can test this PR with a version of ivadomed installed via pypi (as opposed to via github)-- i think i would catch things that we would not catch otherwise-- and the current version on pypi seems too old to test this PR (or is it fine?)

jcohenadad commented on ivadomed/config/config.json:
```diff
@@ -1,22 +1,22 @@
     "command": "train",
-    "gpu": 7,
+    "gpu": 0,
     "log_directory": "spineGeneric",
```
how about something more explicit, like "training_results"?

charleygros replied:
This folder contains: trained model, tensorboard check points (loss and metrics), the configuration file, the dataset split (image paths for training/V/Tessting), but also the nifti files reconstructed from the inference time (testing) and the related csv with metrics for each image.
So this is more a "experiment_directory" than "training_results per se. Would you prefer "experiment_directory" than "log_directory" @jcohenadad ?

jcohenadad replied:
Hum, i see. Fundamentally I would not have a field here, but rather specify output results as a flag when calling ivadomed.

charleygros replied:
Open issue #355

jcohenadad commented:
I'm getting a `NotImplementedError` when running the "Evaluate Model" section of the tutorial, with my pypi installation.

<details><summary>Details</summary>

~~~
(venv-ivadomed) p101317@rosenberg:/scratch$ ivadomed config.json
Log directory already exists: spineGeneric
Using GPU number 0

Selected architecture: Unet, with the following parameters:
        dropout_rate: 0.3
        bn_momentum: 0.9
        depth: 4
        in_channel: 1
        out_channel: 1

Run Inference

Log directory already exists: spineGeneric
Using GPU number 0

Selected transformations for the testing dataset:
        Resample: {'wspace': 0.75, 'hspace': 0.75, 'dspace': 1, 'preprocessing': True}
        CenterCrop: {'size': [128, 128], 'preprocessing': True}
        RandomAffine: {'degrees': 5, 'scale': [0.1, 0.1], 'translate': [0.03, 0.03], 'applied_to': ['im', 'gt']}
        ElasticTransform: {'alpha_range': [28.0, 30.0], 'sigma_range': [3.5, 4.5], 'p': 0.1, 'applied_to': ['im', 'gt']}
        NumpyToTensor: {}
        NormalizeInstance: {'applied_to': ['im']}

Selected architecture: Unet, with the following parameters:
        dropout_rate: 0.3
        bn_momentum: 0.9
        depth: 4
        in_channel: 1
        out_channel: 1
Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:00<00:00, 1105.85it/s]
Loaded 78 axial slices for the testing set.

Loading model: spineGeneric/best_model.pt
Inference - Iteration 0:   0%|                                                                                                                                                   | 0/5 [00:01<?, ?it/s]
Traceback (most recent call last):
  File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/bin/ivadomed", line 8, in <module>
    sys.exit(run_main())
  File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/main.py", line 226, in run_main
    run_main(context)
  File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/main.py", line 216, in run_main
    metric_fns=metric_fns)
  File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/testing.py", line 70, in test
    cuda_available, i_monteCarlo)
  File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/testing.py", line 156, in run_inference
    data_type='gt')
  File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/transforms.py", line 173, in __call__
    sample, metadata = tr.undo_transform(sample, metadata)
  File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/transforms.py", line 82, in undo_transform
    raise NotImplementedError("You need to implement the undo_transform() method.")
NotImplementedError: You need to implement the undo_transform() method.
~~~

</details>

So I guess that answers my comment https://github.com/ivadomed/ivadomed/pull/309#issuecomment-657269465 about the current version on pypi not being compatible with the current branch under test? So should we push the current master version on pypi to be able to test properly?

charleygros commented:
>     * add link to "compatible GPU is available" (section 'run the training') and point to info about compatible GPU, cuda config, etc.

Done in [937ac06](https://github.com/ivadomed/ivadomed/pull/309/commits/937ac0698c056698a366f26fc614abfd6bb747e8)

>     * Pytest failing for py38 #351

I can't reproduce the error on my local machine. Further investigations are needed. This transform is not used in routine: I am suggesting to fix this in another PR.

charleygros commented:
> So I guess that answers my comment [#309 (comment)](https://github.com/ivadomed/ivadomed/pull/309#issuecomment-657269465) about the current version on pypi not being compatible with the current branch under test? So should we push the current master version on pypi to be able to test properly?

Yes absolutely! I just would like to merge these two PRs before (quick fixes), #339 and #342, then new pypi version!

charleygros commented:
> For example, a quick google search led me to this pre-trained model with monai: https://drive.google.com/drive/folders/1HXlYJGvTF3gNGOL0UFBeHVoA6Vh_GqEw

I have modified the table in [f175662](https://github.com/ivadomed/ivadomed/pull/309/commits/f175662646214fd1767326b20ad74d8370534113).
That's likely that a number of the "No" are actually "Yes" :/
The information is not always easily accessible. A solution would be to contact them :D 

> But as you can see i've added a disclaimer "we acknowledge we might be wrong, if that's the case please let us know", to dissuade frustrated researchers from starting a war with us wink

Great! Many thanks, I feel a bit better now :-)
@lrouhier also suggested to add a "Date of last update" (although contained in the commit history) as features are likely to evolve in active repo'.

> also, i've just renamed the column: "Pre-trained models"
:+1:

jcohenadad commented:
> @lrouhier also suggested to add a "Date of last update

Not a big fan if that. We need to constantly update the date as we update the doc. Mistakes will be made.

charleygros commented:
> Not a big fan if that. We need to constantly update the date as we update the doc. Mistakes will be made.

Note: it was only regarding the table comparing the other projects, ie "when did we look at other repo' to complete this table". But fair point.

andreanne-lemay commented:
> I'm getting a `NotImplementedError` when running the "Evaluate Model" section of the tutorial, with my pypi installation.
> 
> Details
> ```
> (venv-ivadomed) p101317@rosenberg:/scratch$ ivadomed config.json
> Log directory already exists: spineGeneric
> Using GPU number 0
> 
> Selected architecture: Unet, with the following parameters:
>         dropout_rate: 0.3
>         bn_momentum: 0.9
>         depth: 4
>         in_channel: 1
>         out_channel: 1
> 
> Run Inference
> 
> Log directory already exists: spineGeneric
> Using GPU number 0
> 
> Selected transformations for the testing dataset:
>         Resample: {'wspace': 0.75, 'hspace': 0.75, 'dspace': 1, 'preprocessing': True}
>         CenterCrop: {'size': [128, 128], 'preprocessing': True}
>         RandomAffine: {'degrees': 5, 'scale': [0.1, 0.1], 'translate': [0.03, 0.03], 'applied_to': ['im', 'gt']}
>         ElasticTransform: {'alpha_range': [28.0, 30.0], 'sigma_range': [3.5, 4.5], 'p': 0.1, 'applied_to': ['im', 'gt']}
>         NumpyToTensor: {}
>         NormalizeInstance: {'applied_to': ['im']}
> 
> Selected architecture: Unet, with the following parameters:
>         dropout_rate: 0.3
>         bn_momentum: 0.9
>         depth: 4
>         in_channel: 1
>         out_channel: 1
> Loading dataset: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 6/6 [00:00<00:00, 1105.85it/s]
> Loaded 78 axial slices for the testing set.
> 
> Loading model: spineGeneric/best_model.pt
> Inference - Iteration 0:   0%|                                                                                                                                                   | 0/5 [00:01<?, ?it/s]
> Traceback (most recent call last):
>   File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/bin/ivadomed", line 8, in <module>
>     sys.exit(run_main())
>   File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/main.py", line 226, in run_main
>     run_main(context)
>   File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/main.py", line 216, in run_main
>     metric_fns=metric_fns)
>   File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/testing.py", line 70, in test
>     cuda_available, i_monteCarlo)
>   File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/testing.py", line 156, in run_inference
>     data_type='gt')
>   File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/transforms.py", line 173, in __call__
>     sample, metadata = tr.undo_transform(sample, metadata)
>   File "/home/GRAMES.POLYMTL.CA/p101317/venv-ivadomed/lib/python3.6/site-packages/ivadomed/transforms.py", line 82, in undo_transform
>     raise NotImplementedError("You need to implement the undo_transform() method.")
> NotImplementedError: You need to implement the undo_transform() method.
> ```
> 
> So I guess that answers my comment [#309 (comment)](https://github.com/ivadomed/ivadomed/pull/309#issuecomment-657269465) about the current version on pypi not being compatible with the current branch under test? So should we push the current master version on pypi to be able to test properly?

The error here is due (I believe) from the fact that `ElasticTransform` don't have an undo_transform. To fix this, in the config file, the `dataset_type` of ElasticTransform can be set to `["training"]`, meaning this data augmentation is only applied during training (which is typically when we want data augmentation). Currently, we can't apply ElasticTransform during test (or eval), because we can't return to the native image by doing an undo of the ElasticTransform. We could force in the ElasticTransform, to be uniquely callable during training, this way it doesn't cause an error.

jcohenadad commented:
> The error here is due (I believe) from the fact that ElasticTransform don't have an undo_transform. To fix this, in the config file, the dataset_type of ElasticTransform can be set to ["training"], meaning this data augmentation is only applied during training (which is typically when we want data augmentation). Currently, we can't apply ElasticTransform during test (or eval), because we can't return to the native image by doing an undo of the ElasticTransform. We could force in the ElasticTransform, to be uniquely callable during training, this way it doesn't cause an error.

so does it mean that @charleygros's suggestion [here](https://github.com/ivadomed/ivadomed/pull/309#issuecomment-657285565) is not valid, and that updating the version on pypi will _not_ solve that issue?

andreanne-lemay commented:
> > The error here is due (I believe) from the fact that ElasticTransform don't have an undo_transform. To fix this, in the config file, the dataset_type of ElasticTransform can be set to ["training"], meaning this data augmentation is only applied during training (which is typically when we want data augmentation). Currently, we can't apply ElasticTransform during test (or eval), because we can't return to the native image by doing an undo of the ElasticTransform. We could force in the ElasticTransform, to be uniquely callable during training, this way it doesn't cause an error.
> 
> so does it mean that @charleygros's suggestion [here](https://github.com/ivadomed/ivadomed/pull/309#issuecomment-657285565) is not valid, and that updating the version on pypi will _not_ solve that issue?

For this error, no, pypi won't solve the issue.

charleygros commented:
> > The error here is due (I believe) from the fact that ElasticTransform don't have an undo_transform. To fix this, in the config file, the dataset_type of ElasticTransform can be set to ["training"], meaning this data augmentation is only applied during training (which is typically when we want data augmentation). Currently, we can't apply ElasticTransform during test (or eval), because we can't return to the native image by doing an undo of the ElasticTransform. We could force in the ElasticTransform, to be uniquely callable during training, this way it doesn't cause an error.
> 
> so does it mean that @charleygros's suggestion [here](https://github.com/ivadomed/ivadomed/pull/309#issuecomment-657285565) is not valid, and that updating the version on pypi will _not_ solve that issue?

What is strange is that you do not get this error with the git installation. My guess is that this disprecancy comes from this PR #335 which deals with transformation at eval time (cf error of @jcohenadad ), which is not included in the pypi release. That's why I suggested this.

charleygros commented:
@jcohenadad ready to be tested with the new release `v2.1.0`.
